Aligning with `UTransport::register_listener()` and `UTransport::unregister_listener()` with `UListener` (this time using instance equality)

PLeVasseur opened the pull request:
Hey there :wave: 

Based on #65, I took my (third!) shot at aligning our `UTransport` trait closer to the spec.

In particular, I have removed the requirement for the implementation to explicitly manage listeners using `String` identifiers, instead leaning on concrete implementations of the `UListener` trait.

This change is to align with [`registerListener()`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#22-registerlistener) and [`unregisterListener()`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#22-registerlistener) by bringing in a [`UListener`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#1-ulistener) trait.

As you read the code, you can see that there was a need for a new wrapper type called `ListenerWrapper` around UListener in order to allow us to disambiguate between _specific instances_ of UListener. The wrapper is also hashable to make it easy to use within containers that require that such as `HashSet` and `HashMap`. The `ListenerWrapper` need only be used in the `up-client-foo-rust` implementation. uEs just impl `UListener`.

Largely resolves the issues I had with the first and second attempts over in #67 and #68 thanks to that wrapper type.

I'm looking to see how this would impact current and future implementations of `UTransport` for UPClients. Feedback would be great from @AnotherDaniel, @sophokles73, @evshary in particular :slightly_smiling_face: 

I [migrated `up-client-zenoh-rust` over to using `UListener`](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) to see how it'd work out if you want to see real-world usage.

PLeVasseur commented:
@sophokles73, @evshary -- could you folks take a look at this one?

I got the [PR for `up-client-zenoh-rust`](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) into better shape too, if you're interested in seeing what this looks like in practice.

PLeVasseur commented on src/ulistener.rs:
```diff
@@ -0,0 +1,29 @@
+    /// # Arguments
+    ///
+    /// * `received` - Either the message or error `UStatus` received
+    fn on_receive(&self, received: Result<UMessage, UStatus>);
```
I think I can see the vision here. :slightly_smiling_face:

WDYT @evshary, @AnotherDaniel ?

PLeVasseur commented on src/ulistener.rs:
```diff
@@ -0,0 +1,29 @@
+/********************************************************************************
```
Yup, makes sense

PLeVasseur replied:
Actually thinking on this one more I could see making a case to keep it separate.

Most end user uEs will not `use up_rust::UTransport` as they will just be using whatever struct `impl`ed it for them.

On the other hand, end users will likely need to `use up_rust::UListener` in order to `impl` it on a struct to use with their `UPClientFoo` of choice.

WDYT?

sophokles73 replied:
That might be true, but still the listener semantically belongs to UTransport. Given that both types are being re-exported, it won't make any difference to client code.

sophokles73 replied:
Are you going to move this to the utransport module?

PLeVasseur commented on src/ulistener.rs:
```diff
@@ -0,0 +1,29 @@
+/// Implementations of `UListener` contain the details for what should occur when a message is received
+/// For more information, please refer to
+/// [uProtocol Specification](https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/up-l1/README.adoc).
+pub trait UListener: Any + Send + Sync {
```
Hokily dokily, I'll have to try this :+1:

PLeVasseur commented on src/listener_wrapper.rs:
```diff
@@ -0,0 +1,69 @@
+use crate::ulistener::UListener;
```
Yup, makes sense

PLeVasseur commented on src/listener_wrapper.rs:
```diff
@@ -0,0 +1,69 @@
+/// In the reverse, if two different instances of `Arc<T>` are used to construct two wrappers,
+/// they will not be considered equivalent due to the Arc pointer being different for each instance
+#[derive(Clone)]
+pub struct ListenerWrapper {
```
Great idea

PLeVasseur commented on src/utransport.rs:
```diff
@@ -85,5 +81,270 @@ pub trait UTransport {
     ///
     /// Returns an error if the listener could not be unregistered, for example if the given listener does not exist.
-    async fn unregister_listener(&self, topic: UUri, listener: &str) -> Result<(), UStatus>;
+    async fn unregister_listener<T>(&self, topic: UUri, listener: &Arc<T>) -> Result<(), UStatus>
```
Yup, this is reasonable

PLeVasseur commented on src/utransport.rs:
```diff
@@ -85,5 +81,270 @@ pub trait UTransport {
+            let identified_listener = ListenerWrapper::new(listener);
+            listeners.insert(identified_listener);
+
+            Err(UStatus::fail_with_code(
+                UCode::OK,
+                format!("{}", listeners.len()),
+            ))
```
You caught me trying to be able to test behavior in a naughty way :laughing: 

I'll have to rethink how to test this

sophokles73 replied:
I do not think that you need to test this explicitly. FMPOV it should be enough to verify that you cannot register the same listener twice and that a message is not handled when no listener is registered ...

PLeVasseur commented on src/listener_wrapper.rs:
```diff
@@ -0,0 +1,69 @@
+
+impl ListenerWrapper {
+    pub fn new<T: UListener>(listener: &Arc<T>) -> Self {
+        let mut hasher = DefaultHasher::new();
+        let pointer = Arc::as_ptr(listener) as *const ();
+        pointer.hash(&mut hasher);
+        let pointer_hash = hasher.finish();
+        let listener = listener.clone();
```
OK, I'll give this a try :slightly_smiling_face:

PLeVasseur commented on src/uri/uresource.rs:
```diff
@@ -13,10 +13,19 @@
 const URESOURCE_ID_VALID_BITMASK: u32 = 0xffff << URESOURCE_ID_LENGTH;
 
+impl Hash for UResource {
+    fn hash<H: Hasher>(&self, state: &mut H) {
```
[Same comment as here](https://github.com/eclipse-uprotocol/up-rust/pull/69#discussion_r1535987855)

PLeVasseur commented:
OK, took all your feedback @sophokles73 except for a couple I had some comments on -- can you take a look again?

sophokles73 commented on src/uri/uentity.rs:
```diff
@@ -13,12 +13,22 @@
 const UENTITY_MAJOR_VERSION_VALID_BITMASK: u32 = 0xffffff << UENTITY_MAJOR_VERSION_LENGTH;
 
+impl Hash for UEntity {
+    fn hash<H: Hasher>(&self, state: &mut H) {
```
IMHO the problem lies with the ambiguity designed into the whole UUri type by means of resolved vs unresolved. It is (still) unclear when a URI needs to be resolved and when it is ok to use an unresolved URI. That said, the problem with your implementation still remains. Can I use an unresolved URI to register a listener?

PLeVasseur replied:
Yeah -- good point.

I have [a PR](https://github.com/eclipse-uprotocol/up-spec/pull/94) going into `up-spec` which clarifies that for `UTransport::send()` the minimum viable UUri is Micro Form.

Perhaps I should also clarify that the same applies to `UTransport::registerListener()` and `UTransport::unregisterListener()`.

Tagging @stevenhartley for his feedback too.

What do you both think?

PLeVasseur commented:
@sophokles73 -- any other open concerns on this one other than the Hash? :slightly_smiling_face:

sophokles73 commented on src/utransport.rs:
```diff
@@ -85,5 +83,327 @@ pub trait UTransport {
+/// Also implements necessary traits to allow hashing, so that you may hold the wrapper type in
+/// collections which require that, such as a `HashMap` or `HashSet`
+///
+/// Note that the implementation is such that if the same instance of `Arc<T>` is used multiple times
```
actually, it is not about the same *instance* of Arc but about the UListener implementation they point to, i.e. it is about the Arc's *pointer*.

sophokles73 commented on src/utransport.rs:
```diff
@@ -85,5 +83,327 @@ pub trait UTransport {
+/// to construct multiple different wrappers, then these wrappers are all considered equivalent
+/// due to using the Arc's pointer as unique identifier
+///
+/// In the reverse, if two different instances of `Arc<T>` are used to construct two wrappers,
+/// they will not be considered equivalent due to the Arc pointer being different for each instance
```
this also needs to be adapted according to the *pointer equality*. Maybe you should simply refer to the `PartialEq::eq` implementation and document there, how Arc<T>s are being compared ...

PLeVasseur replied:
Some good feedback here. I'll try to document this more clearly :+1:

sophokles73 reviewed (commented):
From what I can tell, your Hash implementations are consistent with the PartialEq implementation derived for UAuthority, UEntity and UResource, i.e. if a == b => hash(a) == hash(b). The `PartialEq::eq` implementation will compare *all* of the types' fields while your Hash implementation uses a subset of the fields only. That said, it also means that if two URIs that only use names are not equal, they will still have the same hash code. This is not wrong but it will make operations on HashMaps and HashSets more inefficient because the URIs will end up in the same bucket.

For the time being, it seems to be desirable to be able to use UUri as a key at all, which your implementation allows us to do.

I am still not very happy with the overall ambiguity of the UUri type, but that is a problem out of scope of this PR.

All in all, this LGTM but I would like to check if this approach works with some of my (private) MQTT transport code before merging, though.

PLeVasseur commented:
> For the time being, it seems to be desirable to be able to use UUri as a key at all, which your implementation allows us to do.

Yup, it would be a big help.

> I am still not very happy with the overall ambiguity of the UUri type, but that is a problem out of scope of this PR.

I'd be interested in hearing more about this in an issue on `up-spec` :slightly_smiling_face: 

> All in all, this LGTM but I would like to check if this approach works with some of my (private) MQTT transport code before merging, though.

Sounds like a plan! Can you ping on here with your further feedback?

PLeVasseur commented:
@sophokles73 Updated based on your latest feedback :+1:

sophokles73 reviewed (approved):
LGTM

sophokles73 commented:
@PLeVasseur I have played around with the `UListener` trait in the context of my MQTT transport implementation. I did not run into any problems per se, but I have to admit that I did like the possibility to pass in a closure instead of a Trait object.

Did @evshary test this successfully in up-client-zenoh?

PLeVasseur commented:
> @PLeVasseur I have played around with the `UListener` trait in the context of my MQTT transport implementation. 

:hugs: 

> I did not run into any problems per se, but I have to admit that I did like the possibility to pass in a closure instead of a Trait object.

I agree that passing in a closure does lend itself to more flexibility. On the other hand, I found that if I have a closure which must use some state, this can be clunky, whereas it gets easier to handle using the trait object.

No free lunch, etc :slightly_smiling_face: 

> Did @evshary test this successfully in up-client-zenoh?

I'm not sure. I ended up doing [the leg-work](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) of transitioning `up-client-zenoh-rust` over to using the `UListener` trait, but that PR is going to need to be rebased after @evshary's PR goes in updating to 1.5.7

evshary commented:
> Did @evshary test this successfully in up-client-zenoh?

Not yet. I can do the test on that.

PLeVasseur commented on src/utransport.rs:
```diff
@@ -12,13 +12,37 @@
+/// Implementations of `UListener` contain the details for what should occur when a message is received
+/// For more information, please refer to
+/// [uProtocol Specification](https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/up-l1/README.adoc).
+pub trait UListener: 'static + Send + Sync {
```
- [ ] Based on the feedback today in the uProtocol Zenoh meeting, I need to mark the `on_receive()` and `on_error()` functions as async to allow the `up-client-foo-rust` to asynchronously call these functions so we don't block that thread.

@AnotherDaniel, @sophokles73, @tamarafischer, @stevenhartley -- did I capture the discussion correctly? WDYT?

AnotherDaniel replied:
Yes that sounds like a useful thing to do in this context.

Mallets replied:
I don't believe forcing the trait to be `async` actually solves the problem, although it may mitigate it a bit. If the user still performs long synchronous computation inside the callback, the thread will still be blocked. The choice here is more related to internal implementation and to potentially dedicated a thread pool to the user different from the one used by the middleware.

PLeVasseur replied:
I agree @Mallets, this is just more of enabling the `UTransport` implementation to use a thread pool.

Perhaps we should add some additional comments above `on_receive()` and `on_error()` to clarify that these should be assumed to execute quickly and return.

WDYT?

AnotherDaniel reviewed (commented):
Will only begin to form opinions on usage in the coming weeks, but LGTM so far.

PLeVasseur commented:
> Quick but not rigorous test, it works in UPClientZenoh.

Could you link that here? I'm curious to read :smile:

sophokles73 commented:
So, are we good to merge or do you want to add doc comments to the trait functions first, @PLeVasseur?

PLeVasseur commented:
> So, are we good to merge or do you want to add doc comments to the trait functions first, @PLeVasseur?

IMHO it makes sense for me to:
1. make the trait async and make both `on_receive()` and `on_error()` be async functions
2. add a bit more doc comments above them both clarifying that these functions should finish "fast" and if you have a sync or blocking operation to pass that off within the function

WDYT?

PLeVasseur commented:
Wondering if I've run afoul of something in our CI.

I'm getting an error like this in the latest build:
```
2024-04-03T19:09:27.0641482Z 2024-04-03 19:09:27 [[31mERROR[0m] failed to fetch crates: [1m[91merror[0m: [1mduplicate key `async-std` in table `dev-dependencies`[0m
2024-04-03T19:09:27.0643140Z   [1m[94m-->[0m Cargo.toml:65:1
2024-04-03T19:09:27.0643756Z [1m[94m   |[0m
2024-04-03T19:09:27.0644472Z [1m[94m65 |[0m async-std = { version = "1.12.0" }
2024-04-03T19:09:27.0645414Z [1m[94m   |[0m[1m[91m ^[0m
2024-04-03T19:09:27.0646006Z [1m[94m   |[0m
2024-04-03T19:09:27.0646518Z [0m
2024-04-03T19:09:29.4203855Z 2024-04-03 19:09:27 [[31mERROR[0m] `cargo metadata` exited with an error: [1m[91merror[0m: [1mduplicate key `async-std` in table `dev-dependencies`[0m
2024-04-03T19:09:29.4205599Z   [1m[94m-->[0m Cargo.toml:65:1
2024-04-03T19:09:29.4206246Z [1m[94m   |[0m
2024-04-03T19:09:29.4206668Z [1m[94m65 |[0m async-std = { version = "1.12.0" }
2024-04-03T19:09:29.4207271Z [1m[94m   |[0m[1m[91m ^[0m
2024-04-03T19:09:29.4207686Z [1m[94m   |[0m
2024-04-03T19:09:29.4207972Z [0m
2024-04-03T19:09:29.4225586Z ##[error]Process completed with exit code 1.
```

From what I can tell though, I don't see `async-std` duplicated in the Cargo.toml.

Huh... any thoughts?

PLeVasseur commented:
@sophokles73 -- made some updates, could you take a look now? :slightly_smiling_face: 
1. updated `UListener` to be an async trait
2. updated tests to comply with 1.
3. added doc comments for `on_receive()` and `on_error()` to clarify these should return immediately
4. added a couple of doc examples of impls of `UListener`

PLeVasseur commented on src/utransport.rs:
```diff
@@ -12,13 +12,112 @@
+    /// # Note
+    ///
+    /// `on_receive()` is expected to return almost immediately. If it does not, it could potentially
+    /// block further message receipt. For long-running, synchronous, or blocking operations consider
```
Ahh, you would like a mention of this for implementers of `UTransport`?

I was writing this for the audience of "writers of uEs".

I suppose an implementation of `UTransport` could do one of two things:
1. `await` on `on_receive()`
2. spawn a new async task to run `on_receive()`

In the case of 1, it would matter that `on_recieve()` return fast.

In the case of 2, it becomes a bit more of a gray area on whether it must return fast and may depend on how the `on_receive()` manages any shared mutable state which would prevent forward progress on any future calls to `on_receive()`.

Re-reading what you wrote now, perhaps your point is that since a `UTransport` may do 2, the UListener must be able to be called again immediately for the next incoming message.

Almost seems like to me having two sections of caveats, one for implementers of UListener and another for implementers of UTransport may be helpful.

WDYT?

PLeVasseur commented on src/utransport.rs:
```diff
@@ -85,5 +180,335 @@ pub trait UTransport {
+}
+
+impl ComparableListener {
+    pub fn new(listener: Arc<dyn UListener>) -> Self {
```
Hmm, not sure why I make a clone in here... the function takes ownership of the `Arc<dyn UListener>`. I think that's a leftover. Will remove the clone.

PLeVasseur commented:
@sophokles73 -- Updated based on your feedback. Could you take a look again?

sophokles73 commented on src/utransport.rs:
```diff
@@ -12,13 +12,122 @@
+    ///
+    /// # Note for `UTransport` implementers
+    ///
+    /// Because `on_receive()` is async you may choose to either `.await` it or spawn it onto its
```
you would still need to await it, otherwise no work will be done, if I am not mistaken ...

PLeVasseur replied:
> you would still need to await it, otherwise no work will be done, if I am not mistaken ...

You're 100% right... I didn't word this very well. Let me go clean this up.

sophokles73 commented on src/utransport.rs:
```diff
@@ -58,20 +167,87 @@ pub trait UTransport {
         topic: UUri,
-        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
-    ) -> Result<String, UStatus>;
+        listener: &Arc<dyn UListener>,
```
I do not understand why you have turned this into a *reference to a reference*. Can you elaborate?

sophokles73 commented on src/utransport.rs:
```diff
@@ -238,15 +238,15 @@ pub trait UTransport {
     /// // ...send a clone through when registering...
     /// let my_listener_for_transport = my_listener.clone();
-    /// let register_result = my_transport.register_listener(my_uuri.clone(), &my_listener_for_transport);
+    /// let register_result = my_transport.register_listener(my_uuri.clone(), my_listener_for_transport.clone());
```
`my_listener_for_transport` already is a clone of the original listener ...
FMPOV this should simply be
```rust
let my_listener: Arc<dyn UListener> = Arc::new(MyListener::new());
let register_result = my_transport.register_listener(my_uuri.clone(), my_listener.clone());
// ...and use the original we hung onto when unregistering
let unregister_result = my_transport.unregister_listener(my_uuri, my_listener);
```

PLeVasseur replied:
Ahhh, you're right. Looking at it now it's obvious. Thanks for catching this

sophokles73 reviewed (approved):
LGTM

For the future, I would like to keep updating dependencies separate from functional extensions/improvements (unless required for the implementation). Now that we have Cargo.lock in the repo, we should be more explicit about managing the dependencies.

sophokles73 commented:
@PLeVasseur would you like to squash this, maybe into two commits, the first one for updating the deps and the second one for the rest?

PLeVasseur commented:
> @PLeVasseur would you like to squash this, maybe into two commits, the first one for updating the deps and the second one for the rest?

I like this idea. I need to get better at making conscientious commits when merging PRs. Was a bit too hacky :slightly_smiling_face:  

Went ahead and did this.
